apps/web/modules/ee/contacts/attributes/actions.ts

```diff
@@ -0,0 +1,176 @@
+import { AuthenticatedActionClientCtx } from "@/lib/utils/action-client/types/context";
+import {
+  getOrganizationIdFromEnvironmentId,
+  getProjectIdFromEnvironmentId,
+} from "@/lib/utils/helper";
+import { isSafeIdentifier } from "@/lib/utils/safe-identifier";
+import { withAuditLogging } from "@/modules/ee/audit-logs/lib/handler";
+import {
+  createContactAttributeKey,
+  updateContactAttributeKey,
+  deleteContactAttributeKey,
+  getContactAttributeKeyById,
+} from "@/modules/ee/contacts/lib/contact-attribute-keys";
+
+const ZCreateContactAttributeKeyAction = z.object({
+  environmentId: ZId,
+  key: z.string().refine(
+    (val) => isSafeIdentifier(val),
+    {
+      message:
+        "Key must be a safe identifier: only lowercase letters, numbers, and underscores, and must start with a letter",
+    }
+  ),
+  name: z.string().optional(),
+  description: z.string().optional(),
+});
+
+export const createContactAttributeKeyAction = authenticatedActionClient
+  .schema(ZCreateContactAttributeKeyAction)
+  .action(
+    withAuditLogging(
+      "created",
+      "contactAttributeKey",
+      async ({ ctx, parsedInput }: { ctx: AuthenticatedActionClientCtx; parsedInput: Record<string, any> }) => {
+        const organizationId = await getOrganizationIdFromEnvironmentId(parsedInput.environmentId);
+        const projectId = await getProjectIdFromEnvironmentId(parsedInput.environmentId);
+
+        await checkAuthorizationUpdated({
+          userId: ctx.user.id,
+          organizationId,
+          access: [
+            {
+              type: "organization",
+              roles: ["owner", "manager"],
+            },
+            {
+              type: "projectTeam",
+              minPermission: "readWrite",
+              projectId,
+            },
+          ],
+        });
+
+        ctx.auditLoggingCtx.organizationId = organizationId;
+
+        const contactAttributeKey = await createContactAttributeKey({
+          environmentId: parsedInput.environmentId,
+          key: parsedInput.key,
+          name: parsedInput.name,
+          description: parsedInput.description,
+        });
+
+        ctx.auditLoggingCtx.newObject = contactAttributeKey;
+
+        return contactAttributeKey;
+      }
+    )
+  );
+
+const ZUpdateContactAttributeKeyAction = z.object({
+  id: ZId,
+  name: z.string().optional(),
+  description: z.string().optional(),
+});
+
+export const updateContactAttributeKeyAction = authenticatedActionClient
+  .schema(ZUpdateContactAttributeKeyAction)
+  .action(
+    withAuditLogging(
+      "updated",
+      "contactAttributeKey",
+      async ({ ctx, parsedInput }: { ctx: AuthenticatedActionClientCtx; parsedInput: Record<string, any> }) => {
+        // Fetch existing key to check authorization and get environmentId
+        const existingKey = await getContactAttributeKeyById(parsedInput.id);
+
+        if (!existingKey) {
+          throw new Error("Contact attribute key not found");
```

Comment:
When we throw the error from the action it gets interpreted as a server error and the generic error message `Something went wrong` is shown in the toast, so the user is never able to see this message. I think a better approach would be to use the `Result` type (present in the error-handler types in the types package) and return meaningful error messages instead of throwing them. This way we get actual error messages to show to the user on the frontend